spec/controllers/users/totp_setup_controller_spec.rb

```diff
@@ -41,64 +41,59 @@
     context 'when user presents invalid code' do
       before do
         sign_in_as_user
+
+        stub_analytics
+        allow(@analytics).to receive(:track_event)
+
         get :new
         patch :confirm, code: 123
       end
 
-      it 'redirects back to authenticator_setup_path' do
+      it 'redirects with an error message' do
         expect(response).to redirect_to(authenticator_setup_path)
-      end
-
-      it 'sets flash[:error] message' do
         expect(flash[:error]).to eq t('errors.invalid_totp')
-      end
-
-      it 'does not enable TOTP for the current user' do
```

Comment:
Indeed, I think that can be achieves with "before :all" rather than combining cases into one.